### PR TITLE
Add websocket server

### PR DESCRIPTION
**Pull request by Munikumar09**
## Summary by CodeRabbit

* **New Features**
  * Added asynchronous Redis-based stock data caching with timestamp-aware updates.
  * Introduced a singleton FastAPI app managing lifecycle events, real-time WebSocket server, Kafka consumer, and Redis connection.
  * Developed WebSocket server protocol and connection manager for efficient client subscriptions and broadcasting.
  * Added thread-safe singleton metaclass and asyncio utilities for event loop and coroutine management.
  * Provided new Kafka consumer and producer implementations with robust error handling and backoff.
  * Added detailed health check and improved server startup/shutdown logic.
  * Introduced a WebSocket client script for subscribing to stock updates.

* **Refactor**
  * Migrated Kafka clients from `kafka-python` to `confluent-kafka`.
  * Replaced generic streaming abstractions with explicit Producer and Consumer interfaces.
  * Centralized Redis connection management into singleton classes for sync and async clients.
  * Updated environment variable fetching to support defaults and optional values.
  * Refactored data savers to use polling-based Kafka consumption with enhanced error handling.
  * Improved database CRUD utilities with better type validation and primary key handling.

* **Bug Fixes**
  * Fixed input validation and timestamp parsing in stock cache updates.
  * Resolved issues with CSV, JSONL, and SQLite savers handling Unicode, nulls, and special characters.
  * Corrected environment variable usage and configuration defaults in multiple modules.

* **Chores**
  * Updated Docker Compose files to assign Kafka and Redis containers to user-defined networks.
  * Added a startup shell script orchestrating WebSocket client and server with health checks.
  * Enhanced pre-commit hooks to skip deleted or missing Python files.
  * Added new secret environment variables to CI workflows for Kafka configuration.

* **Documentation**
  * Added comprehensive docstrings and logging across new and refactored modules.
  * Improved test documentation and organization for clarity and maintainability.

* **Style**
  * Standardized function signatures and added explicit type annotations in tests and utilities.
  * Renamed test classes and fixtures for consistency.

* **Tests**
  * Added extensive new and refactored tests covering Kafka consumer/producer, WebSocket server and protocol, instrument caching, asyncio utilities, and data savers.
  * Removed deprecated streaming tests and legacy modules.
  * Expanded integration and edge case tests for database CRUD, caching, and streaming components.

* **Revert**
  * Removed legacy KafkaStreamer and Streamer abstractions, replaced by updated Producer/Consumer implementations and tests.

### CODE REVIEW

File: backend/pyproject.toml [hunk 1]
```diff
@@ -25,7 +25,8 @@ numpy = "^1.26.4"
 pandas = "^2.2.2"
 sqlmodel = "^0.0.19"
 registrable = "^0.0.4"
-kafka-python-ng = "^2.2.2"
+aiokafka = "^0.10.0" # Added for async Kafka consumption
+redis = {extras = ["hiredis"], version = "^5.0.1"} # Added for caching
```
[the remark this answers is not in the document]
Comment (Munikumar09) [reply backend/pyproject.toml:29]: Done
